icu4c/source/test/cintltst/custrtst.c

```diff
@@ -1124,6 +1128,32 @@ TestUnescape() {
     /* ### TODO: test u_unescapeAt() */
 }
 
+static void
+TestUnescapeRepeatedSurrogateLead20725() {
+    const int32_t repeat = 20000;
+    const int32_t srclen = repeat * 6 + 1;
+    char *src = (char*)malloc(srclen);
+    UChar *dest = (UChar*) malloc(sizeof(UChar) * (repeat + 1));
+    if (src == NULL || dest == NULL) {
+        log_err("memory allocation error");
+    }
+    for (int32_t i = 0; i < repeat; i++) {
+      uprv_strcpy(src + (i * 6), "\\ud841");
+    }
+    int32_t len = u_unescape(src, dest, repeat);
+    if (len != repeat) {
+        log_err("failure in u_unescape()");
+    }
+    for (int32_t i = 0; i < repeat; i++) {
+      if (dest[i] != 0xd841) {
+        log_err("failure in u_unescape() return value");
+      }
+    }
+    free(src);
+    free(dest);
+
```

Comment:
Please add a few simple test cases to make sure that the code recovers properly. For example:
*   `"\\ud841\\x5A"` --> `u"\ud841Z"`
*   `"\\ud841\\U00050005"` --> `u"\ud841\U00050005"`
*   `"\\ud841\\xXX"` --> `u"\ud841\\xXX"`

The last one is ill-formed. I am just guessing at the output for that one.
Hopefully UTF-16 string literals -- u"..." -- work in C test code on our platforms.

Reply:
ill-formed u"" literal does not work in C, clang give me a "invalid universal character" error
```
custrtst.c:1155:26: error: invalid universal character
    if (u_strcmp(dest, u"\ud841Z")!=0) {
                         ^~~~~~
```